core/controllers/base_test.py

```diff
@@ -628,16 +583,113 @@ def test_unauthorized_user_exception_raised_when_session_is_invalid(self):
                 auth_services, 'get_auth_claims_from_request',
                 error=auth_domain.InvalidAuthSessionError('uh-oh')))
 
-            response = self.testapp.get('/', expect_errors=True)
+            response = self.get_html_response('/', expected_status_int=401)
 
         self.assert_matches_regexps(logs, [
             r'User session is invalid!',
-            r'User must sign in again\nTraceback \(most recent call last\):\n',
+            r'Traceback \(most recent call last\):.*InvalidAuthSessionError',
+            r'Exception raised:',
         ])
-        self.assertEqual(response.status_int, 500)
         self.assertEqual(call_counter.times_called, 1)
 
 
+class MaintenanceModeTests(test_utils.GenericTestBase):
+    """Tests BaseHandler behavior when maintenance mode is enabled.
+
+    Each test case runs within a context where ENABLE_MAINTENANCE_MODE is True.
+    """
+
+    def setUp(self):
+        super(MaintenanceModeTests, self).setUp()
+        with contextlib2.ExitStack() as exit_stack:
+            exit_stack.enter_context(
+                self.swap(feconf, 'ENABLE_MAINTENANCE_MODE', True))
+            self.exit_stack = exit_stack.pop_all()
+
+    def tearDown(self):
+        self.exit_stack.close()
+        super(MaintenanceModeTests, self).tearDown()
+
+    def test_html_response_is_rejected(self):
+        destroy_auth_session_call_counter = self.exit_stack.enter_context(
+            self.swap_with_call_counter(auth_services, 'destroy_auth_session'))
+
+        response = self.get_html_response(
+            '/community-library', expected_status_int=503)
+
+        self.assertIn('<maintenance-page>', response.body)
+        self.assertNotIn('<library-page>', response.body)
+        self.assertEqual(destroy_auth_session_call_counter.times_called, 1)
+
+    def test_html_response_is_not_rejected_when_user_is_super_admin(self):
+        self.exit_stack.enter_context(self.super_admin_context())
```

Comment:
Can exit_stack be renamed to a more neutral name like context_stack? It seems to be the context for the overall test, and not just something applied on-exit.

Reply:
Done.